vignettes/articles/managing-tokens-securely.Rmd

```diff
@@ -0,0 +1,198 @@
+
+One way to make the `secret_*()` functions available for local experimentation is to call `devtools::load_all()`, which exposes all internal objects in a package:
+
+```{r eval = FALSE}
+devtools::load_all("path/to/source/of/gargle/")
+```
+
+The approach I'll take in this article is to call these functions via `:::`.
+
+We go ahead and attach gargle in the normal way, which makes the exported functions readily available.
+
+```{r setup}
+library(gargle)
+```
+
+## Overview of the approach
+
+  * Generate a random PASSWORD and give it a self-documenting name, e.g. `GARGLE_PASSWORD`.
+    Store as an environment variable.
+  * Identify a secret file of interest, such as the JSON representing a
+    service account token. This is presumably stored *outside* your package.
+  * Use the PASSWORD to apply a method for symmetric encryption to the target
+    file.
+  * Store the encrypted file in a predictable place *within* your package.
+  * Store or pass the PASSWORD as an environment variable everywhere you'll need
+    to decrypt the secret.
+    - Check that the platform has support for keeping the PASSWORD concealed.
+    - Make sure you don't do anything in your own code that would dump it to,
+      e.g., a log file.
+  * Rig your tests to determine if the key is available and, therefore,
+    whether decryption is going to be possible.
+    - If "no", carry on gracefully with any tests that don't require auth.
+    - If "yes", decrypt the secret and put the associated token into force
+      globally for the test run or on an "as needed" basis in individual tests.
+
+## Annotated code-through
+
+### Generate a name for the PASSWORD
+
+We believe this should record the associated package. `secret_pw_name()` does this.
+
+```{r}
+(pw_name <- gargle:::secret_pw_name("gargle"))
+```
+
+### Generate a random PASSWORD
+
+In real life, you should keep the output of `secret_pw_gen()` to yourself! We reveal it here as part of the exposition.
+
+```{r}
+(pw <- gargle:::secret_pw_gen())
+```
+
+### Define env var in `.Renviron`
+
+Combine the name and value to form a line like this in your user-level `.Renviron` file:
+
+```{r, echo = FALSE, results = "asis"}
+cat(paste0(pw_name, "=", pw), sep = "\n")
+```
+
+`usethis::edit_r_environ()` can help create or open this file. We recommend using the user-level `.Renviron`, as opposed to project-level, for the usual reason  that this makes it less likely you will share sensitive information by mistake. Make sure `.Renviron` ends in a newline; the lack of this is a notorious cause of silent failure. Remember you'll need to restart R for the newly defined environment variable to take effect.
```

Comment:
You quite possibly have good reasons for not mentioning it but you can also reload .Renviron with `readRenviron("~/.Renviron")`

Reply:
😲 I didn't know about that function!